daemon/logger/local/read.go

```diff
@@ -99,63 +98,71 @@ func getTailReader(ctx context.Context, r loggerutils.SizeReaderAt, req int) (io
 type decoder struct {
 	rdr   io.Reader
 	proto *logdriver.LogEntry
-	buf   []byte
+	// buf keeps bytes from rdr.
+	buf []byte
+	// offset is the position in buf.
+	// If offset > 0, buf[offset:] has bytes which are read but haven't used.
+	offset int
+	// nextMsgLen is the length of the next log message.
+	// If nextMsgLen = 0, a new value must be read from rdr.
+	nextMsgLen int
 }
 
-func (d *decoder) Decode() (*logger.Message, error) {
-	if d.proto == nil {
-		d.proto = &logdriver.LogEntry{}
-	} else {
-		resetProto(d.proto)
-	}
-	if d.buf == nil {
-		d.buf = make([]byte, initialBufSize)
-	}
-	var (
-		read int
-		err  error
-	)
-
+func (d *decoder) readRecord(size int) error {
+	var err error
 	for i := 0; i < maxDecodeRetry; i++ {
 		var n int
-		n, err = io.ReadFull(d.rdr, d.buf[read:encodeBinaryLen])
+		n, err = io.ReadFull(d.rdr, d.buf[d.offset:size])
+		d.offset += n
```

Comment:
It seems like we are only using offset inside this function to handle `ErrUnexpectedEOF`. Do we need to store it on the object at all?

Reply:
Yes. readRecord() have no guarantees about where it ends. It would be the middle of the log message. So the decoder has to keep the fact and read the remaining in that case.